File: packages/database/src/repositories/billing.repository.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";

const inv = vi.hoisted(() => ({
  findFirst: vi.fn(),
  findUnique: vi.fn(),
  findMany: vi.fn(),
  create: vi.fn(),
  count: vi.fn(),
  updateMany: vi.fn(),
}));

vi.mock("../client", () => ({ prisma: { membershipInvoice: inv } }));

import { createInvoice, findByIdForUser, listForUser } from "./billing.repository";

const D = (iso: string) => new Date(iso);
const dec = (n: number) => ({ toNumber: () => n });

function row(overrides: Record<string, unknown> = {}) {
  return {
    id: "inv-1",
    receiptNo: "BIKIE-2026-000006",
    userId: "user-1",
    accountType: "RIDER",
    userMembershipId: "um-1",
    partnerMembershipId: null,
    planId: "plan-1",
    planName: "Membership",
    amount: dec(99),
    currency: "INR",
    durationDays: 365,
    membershipStartDate: D("2026-08-30T00:00:00.000Z"),
    membershipEndDate: D("2027-08-30T00:00:00.000Z"),
    customerName: "Priya Verma",
    customerPhone: "+919876543210",
    status: "PAID",
    paymentId: "pay_abc",
    razorpayPaymentId: "pay_abc",
    razorpayOrderId: "order_abc",
    confirmationSmsSentAt: null,
    createdAt: D("2026-08-30T09:00:00.000Z"),
    paidAt: D("2026-08-30T09:00:00.000Z"),
    ...overrides,
  };
}

const baseInput = {
  userId: "user-1",
  accountType: "RIDER" as const,
  userMembershipId: "um-1",
  planId: "plan-1",
  planName: "Membership",
  amount: 99,
  durationDays: 365,
  membershipStartDate: D("2026-08-30T00:00:00.000Z"),
  membershipEndDate: D("2027-08-30T00:00:00.000Z"),
  customerName: "Priya Verma",
  customerPhone: "+919876543210",
  paymentId: "pay_abc",
  razorpayPaymentId: "pay_abc",
  razorpayOrderId: "order_abc",
  paidAt: D("2026-08-30T09:00:00.000Z"),
};

afterEach(() => vi.clearAllMocks());

describe("billing.repository.createInvoice — ADR-070", () => {
  it("mints a new receipt as BIKIE-<year>-<6 digits> and writes the full snapshot", async () => {
    inv.findFirst.mockResolvedValueOnce(null); // no existing
    inv.count.mockResolvedValueOnce(5); // 5 used this year → next is 000006
    inv.create.mockImplementationOnce(async ({ data }: { data: Record<string, unknown> }) => row({ receiptNo: data.receiptNo }));

    const result = await createInvoice(baseInput);

    const createArg = inv.create.mock.calls[0][0].data;
    expect(createArg.receiptNo).toBe(`BIKIE-${new Date().getFullYear()}-000006`);
    expect(createArg).toMatchObject({
      userId: "user-1",
      accountType: "RIDER",
      userMembershipId: "um-1",
      planName: "Membership",
      amount: 99,
      currency: "INR",
      durationDays: 365,
      razorpayPaymentId: "pay_abc",
    });
    expect(result.receiptNo).toBe(`BIKIE-${new Date().getFullYear()}-000006`);
    expect(result.amount).toBe(99);
  });

  it("is idempotent: an existing invoice for the same membership row is returned, no create", async () => {
    inv.findFirst.mockResolvedValueOnce(row());

    const result = await createInvoice(baseInput);

    expect(inv.create).not.toHaveBeenCalled();
    expect(result.id).toBe("inv-1");
    expect(result.receiptNo).toBe("BIKIE-2026-000006");
  });

  it("recovers from a P2002 race by returning the invoice the winner created", async () => {
    inv.findFirst
      .mockResolvedValueOnce(null) // pre-check: none
      .mockResolvedValueOnce(row()); // post-P2002: winner's row
    inv.count.mockResolvedValueOnce(0);
    inv.create.mockRejectedValueOnce(Object.assign(new Error("dup"), { code: "P2002" }));

    const result = await createInvoice(baseInput);

    expect(result.id).toBe("inv-1");
  });

  it("snapshot is immutable — reads return the stored amount even if the plan later changes", async () => {
    // Invoice was written at ₹99; the plan is now ₹149. listForUser reads the STORED row.
    inv.findMany.mockResolvedValueOnce([row({ amount: dec(99) })]);

    const [summary] = await listForUser("user-1");

    expect(summary.amount).toBe(99); // not 149
  });
});

describe("billing.repository.findByIdForUser — ownership (IDOR)", () => {
  it("returns the invoice for its owner", async () => {
    inv.findUnique.mockResolvedValueOnce(row({ userId: "user-1" }));
    const result = await findByIdForUser("inv-1", "user-1");
    expect(result?.id).toBe("inv-1");
  });

  it("returns null when the invoice belongs to another user (no existence leak)", async () => {
    inv.findUnique.mockResolvedValueOnce(row({ userId: "someone-else" }));
    const result = await findByIdForUser("inv-1", "user-1");
    expect(result).toBeNull();
  });

  it("returns null when the invoice does not exist", async () => {
    inv.findUnique.mockResolvedValueOnce(null);
    const result = await findByIdForUser("nope", "user-1");
    expect(result).toBeNull();
  });
});

File: apps/mobile/test/features/billing/billing_repository_test.dart
import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/core/network/api_exception.dart';
import 'package:mobile/features/billing/data/billing_repository.dart';
import 'package:mocktail/mocktail.dart';

class _MockDio extends Mock implements Dio {}

void main() {
  setUpAll(() {
    registerFallbackValue(RequestOptions(path: '/api/billing/history'));
  });

  late _MockDio dio;
  late BillingRepository repository;

  setUp(() {
    dio = _MockDio();
    repository = BillingRepository(dio);
  });

  Response<T> ok<T>(String path, T data) =>
      Response<T>(requestOptions: RequestOptions(path: path), statusCode: 200, data: data);

  Map<String, dynamic> summaryJson({String id = 'inv-1'}) => {
        'id': id,
        'receiptNo': 'BIKIE-2026-000001',
        'accountType': 'RIDER',
        'planName': 'Membership',
        'amount': 99,
        'currency': 'INR',
        'status': 'PAID',
        'paidAt': '2026-08-30T09:00:00.000Z',
        'membershipStartDate': '2026-08-30T00:00:00.000Z',
        'membershipEndDate': '2027-08-30T00:00:00.000Z',
      };

  group('BillingRepository.getHistory', () {
    test('gets /api/billing/history and parses the invoices list', () async {
      when(() => dio.get('/api/billing/history')).thenAnswer(
        (_) async => ok('/api/billing/history', {
          'invoices': [summaryJson(id: 'inv-1'), summaryJson(id: 'inv-2')],
        }),
      );

      final invoices = await repository.getHistory();

      expect(invoices, hasLength(2));
      expect(invoices.first.receiptNo, 'BIKIE-2026-000001');
      expect(invoices.first.amount, 99);
    });

    test('returns an empty list when the envelope has no invoices key', () async {
      when(() => dio.get('/api/billing/history'))
          .thenAnswer((_) async => ok('/api/billing/history', <String, dynamic>{}));

      expect(await repository.getHistory(), isEmpty);
    });
  });

  group('BillingRepository.getInvoice', () {
    test('gets /api/billing/invoices/:id and parses the detail envelope', () async {
      when(() => dio.get('/api/billing/invoices/inv-1')).thenAnswer(
        (_) async => ok('/api/billing/invoices/inv-1', {
          'invoice': {
            ...summaryJson(),
            'userId': 'user-1',
            'customerName': 'Priya Verma',
            'customerPhone': '+919876543210',
            'planId': 'plan-1',
            'durationDays': 365,
            'razorpayPaymentId': 'pay_abc',
            'razorpayOrderId': 'order_abc',
            'paymentId': 'pay_abc',
            'confirmationSmsSentAt': null,
            'createdAt': '2026-08-30T09:00:00.000Z',
          },
        }),
      );

      final inv = await repository.getInvoice('inv-1');

      expect(inv.receiptNo, 'BIKIE-2026-000001');
      expect(inv.customerName, 'Priya Verma');
      expect(inv.durationDays, 365);
      expect(inv.summary.amount, 99);
    });

    test('surfaces a 404 (someone else\'s / missing invoice) as a typed ApiException', () async {
      when(() => dio.get('/api/billing/invoices/nope')).thenThrow(
        DioException(
          requestOptions: RequestOptions(path: '/api/billing/invoices/nope'),
          response: Response(
            requestOptions: RequestOptions(path: '/api/billing/invoices/nope'),
            statusCode: 404,
            data: {'error': 'NOT_FOUND'},
          ),
        ),
      );

      expect(() => repository.getInvoice('nope'), throwsA(isA<ApiException>()));
    });
  });
}
